File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/diffusion_cameraman_tutorial.m
close all; clear all; clc; 

% diffusionTutorial - A from-scratch implementation of a simple diffusion model.
% This tutorial uses the built-in "cameraman.tif" image as the input data.

%% Load and Preprocess the Image
I = imread('data/cameraman.tif');
I = im2double(I);           % Normalize image to [0, 1]
[H, W] = size(I);
imgSize = [H, W];
x0 = I(:);                  % Flatten image into a vector (H*W x 1)

%% Diffusion Model Parameters
T = 50;  % Number of diffusion steps
% Define a linear noise schedule: beta increases linearly from 0.0001 to 0.02
beta = linspace(0.0001, 0.02, T);
alpha = 1 - beta;

% Pre-allocate matrices to store forward process and noise
x_forward = zeros(length(x0), T+1);
x_forward(:,1) = x0;
eps_store = zeros(length(x0), T);  % to store noise for each step

%% Forward Diffusion Process
% Iteratively add noise to the image
for t = 1:T
    eps_t = randn(size(x0));
    eps_store(:,t) = eps_t;
    x_forward(:,t+1) = sqrt(alpha(t)) * x_forward(:,t) + sqrt(beta(t)) * eps_t;
end

%% Reverse Diffusion Process
% Use the stored noise to reverse the forward process
x_reverse = zeros(size(x_forward));
x_reverse(:,T+1) = x_forward(:,T+1);  % Start with the final noisy image
for t = T:-1:1
    % Reverse update: x_{t-1} = (x_t - sqrt(beta(t))*eps_t)/sqrt(alpha(t))
    x_reverse(:,t) = (x_reverse(:,t+1) - sqrt(beta(t)) * eps_store(:,t)) / sqrt(alpha(t));
end

% Reshape the final reconstructed vector to image format
I_recon = reshape(x_reverse(:,1), H, W);
I_noisy = reshape(x_forward(:,T+1), H, W);

%% Visualization
figure;
subplot(1,3,1);
    imshow(I);
    title('Original Image');
subplot(1,3,2);
    imshow(I_noisy);
    title({'Noisy Image';'(Forward Process)'});
subplot(1,3,3);
    imshow(I_recon);
    title({'Reconstructed Image';'(Reverse Process)'});

% save_all_figs_OPTION('results/diffusion2','png',1)

% Plot absolute difference between original and reconstructed images
figure;
imagesc(abs(I - I_recon));
colormap gray;
colorbar;
title('Absolute Difference: Original vs. Reconstructed');

% save_all_figs_OPTION('results/diffusion3','png',1)

File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/diffusion_cameraman_tutorial.py
#!/usr/bin/env python3
"""
diffusionTutorial_cameraman.py
------------------------------
A from-scratch implementation of a simple diffusion model applied to the "cameraman" image.
The forward process iteratively adds noise to the image; the reverse process uses the stored noise to reconstruct the image.
"""

import numpy as np
import matplotlib.pyplot as plt
from skimage import data, img_as_float, transform

# 1. Load and preprocess image
I = img_as_float(data.camera())
H, W = I.shape
x0 = I.flatten()

# 2. Diffusion model parameters
T = 50  # number of diffusion steps
beta = np.linspace(0.0001, 0.02, T)  # noise schedule
alpha = 1 - beta

x_forward = np.zeros((x0.size, T+1))
x_forward[:, 0] = x0
eps_store = np.zeros((x0.size, T))

# 3. Forward Diffusion Process
for t in range(T):
    eps_t = np.random.randn(x0.size)
    eps_store[:, t] = eps_t
    x_forward[:, t+1] = np.sqrt(alpha[t]) * x_forward[:, t] + np.sqrt(beta[t]) * eps_t

# 4. Reverse Diffusion Process
x_reverse = np.zeros_like(x_forward)
x_reverse[:, T] = x_forward[:, T]
for t in range(T-1, -1, -1):
    x_reverse[:, t] = (x_reverse[:, t+1] - np.sqrt(beta[t]) * eps_store[:, t]) / np.sqrt(alpha[t])

I_recon = x_reverse[:, 0].reshape(H, W)
I_noisy = x_forward[:, T].reshape(H, W)

# 5. Visualization
plt.figure(figsize=(12,4))
plt.subplot(1,3,1)
plt.imshow(I, cmap='gray')
plt.title("Original Image")
plt.axis('off')

plt.subplot(1,3,2)
plt.imshow(I_noisy, cmap='gray')
plt.title("Noisy Image (Forward Process)")
plt.axis('off')

plt.subplot(1,3,3)
plt.imshow(I_recon, cmap='gray')
plt.title("Reconstructed Image (Reverse Process)")
plt.axis('off')
plt.show()

plt.figure()
plt.imshow(np.abs(I - I_recon), cmap='gray')
plt.colorbar()
plt.title("Absolute Difference: Original vs. Reconstructed")
plt.show()
